### PR TITLE
UIA: Support half checked state for checkboxes and toggle buttons

### LINKED ISSUE #13975: NVDA does not support the ToggleState_Indeterminate for a three-state checkbox that implements the UIATogglePattern
### Steps to reproduce:
1. create a new WPF application.
2. Add the following code to the XMAL file to create a three state checkbox: `` <CheckBox Content="Three state" IsChecked="False" IsThreeState="True"></CheckBox> ``
3. Run the application and set focus to the checkbox.
4. Use the space bar to toggle the state of the checkbox. It is possible to confirm the state at any time by opening the NVDA console with NVDA +CTRL +Z and entering the following python code: `` focus.UIATogglePattern.CurrentToggleState ``. The [partially checked](https://docs.microsoft.com/en-us/windows/win32/api/uiautomationcore/ne-uiautomationcore-togglestate) state corresponds to state 2 in the output.

### Actual behavior:

The states are reported as follows:
* Three state  check box  not checked
* checked
* not checked

### Expected behavior:

The states must be reported as follows:
* Three state  check box  not checked
* checked
* partially checked

the third state which is now reported as not checked must map to an NVDA halfchecked state instead.

### NVDA logs, crash dumps and other attachments:
Not applicable.
#### NVDA installed/portable/running from source:
Installed.
#### NVDA version:
2022.2
#### Windows version:
Edition	Windows 11 Enterprise Insider Preview

Version	22H2

OS build	25163.1000

#### Name and version of other software in use when reproducing the issue:
Not applicable.
#### Does the issue still occur after restarting your computer?
Yes
#### Have you tried any other versions of NVDA? If so, please report their behaviors.
No
#### If NVDA add-ons are disabled, is your problem still occurring?
Yes
#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?
Not applicable

### PR DESCRIPTION
**Pull request by mzanm**
### Link to issue number:
Fixes #13975 
### Summary of the issue:
NVDA does not support 3 state checkboxes and toggle buttons with UIA. When using WPF or Microsoft edge UIA.
### Description of user facing changes
When a checkbox is partially checked, NVDA will report half checked, same as the turn features control panel applet.
### Description of development approach
in NVDAObjects\UIA, if toggle state is indeterminate, add half checked state
### Testing strategy:
Tested using WPF three-state checkbox and toggle button. And with using UIA chromium-baste browsers on, tested with Microsoft Edge.
### Known issues with pull request:
None.

### Change log entries:
Bug fixes
- In programs using UI Automation, Partially checked checkboxes will be reported correctly.

### Code Review Checklist:

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English

### CODE REVIEW

**Comment by seanbudd** [comment]
This LGTM, is there a reason it is still a draft?

**Comment by mazen428** [comment]
It currently says for toggle buttons
Toggle button not pressed half checked
apart from the negative state bug, I'm asking if that's ok as is or a new state would be better to be consistent with pressed and not pressed of toggle  buttons.

**Comment by feerrenrut** [comment]
Please ensure that you test with how this affects the reporting of "indeterminate progress bars" which may use the same state, but a different role.

We don't want them to be reported as "progress bar  half checked".

See implementation of "busy indicator": https://github.com/nvaccess/nvda/pull/13428
and issues:
- https://github.com/nvaccess/nvda/issues/10644
- https://github.com/nvaccess/nvda/issues/13910

**Comment by mazen428** [comment]
This does not effect progress bars as the indeterminate state is fetched using the UIA toggle pattern, which is either on, off, or indeterminate. It's specific to toggle buttons and checkboxes. UIA doesn't seem to have a way to fetch if a progress bar is indeterminate.

**Comment by rperez030** [comment]
Half pressed for toggle sounds good to me. I cannot tell if it conveys the way it looks visually, but conveys the meaning.

**Comment by OzancanKaratas** [comment]
@mazen428, please mark this pull request as ready for review.

**Comment by mazen428** [comment]
I don't have a braille display so can't add or test the state in braille, so if someone could that'd be great

**Comment by OzancanKaratas** [comment]
@cary-rowen, would you test this on your braille display?

**Comment by cary-rowen** [comment]
Answer what you care about first:
Yes, I tested with Seika Notetaker, the checkbox can render three states in braille.

The sample I tested is as follows(Enchantment table checkBox):
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/checkbox

**Comment by cary-rowen** [comment]
Also, I don't know if it will be off topic.
Still using the sample given above, But with UIA enabled NVDA doesn't report checkbox checked status in time.

NVDA can report the checked state of a checkbox in time when UIA is disabled, but Braille can only have two states, checked or unchecked.

**Comment by mazen428** [comment]
Thank you. Could you test with braille using this WPF application?
You might need the dot net 6 runtime to run it, simply unpack the archive and you can find the exe inside the bin folder.
the app just have 3 state toggle button and a 3 state checkbox.
[wpftest.zip](https://github.com/nvaccess/nvda/files/9332412/wpftest.zip)
Alternatively, the code is included so you could run it if you have the SDK.

**Comment by rperez030** [comment]
WPF example:
The checkbox works as expected for both speech and Braille. the indeterminate state is represented in Braille with the appropriate symbol for the halfchecked state.
The toggle button reports the three states as expected in speech, but the halfpressed state is represented in braille with the whole state instead of the symbol (half pressed).
Edge with the UIA flag set to Yes:
checkboxes work as expected for both speech and braille. The half checked checkbox is represented in braille with the appropriate symbol for the halfchecked state.
I also observed that NVDA fails to report the state change for the checkboxes with UIA enabled, but I believe that should be handled outside this PR.

**Comment by mazen428** [comment]
Ok, I think the half pressed braille state is fixed.
By NVDA not announcing UIA checkbox state, is that with WPF or just chromium UIA?

**Comment by cary-rowen** [comment]
> By NVDA not announcing UIA checkbox state, is that with WPF or just chromium UIA?
The issue only exists in Edge's web pages.
If you want to reproduce it, you must enable NVDA's UIA support for Edge in Advanced Settings Panel.

**Comment by mazen428** [comment]
I can reproduce this on 2022.2, I'm not sure how to fix it and I agree it should be fixed in another PR.
I've also noticed if you go into focus mode it reports the state changes, so it might be a UIA browse mode state change issue, so it's unrelated to what's done here.

**Comment by rperez030** [comment]
> Ok, I think the half pressed braille state is fixed.

Yes it is. Thanks @mazen428  for your work on this PR. It has also helped me in my journey to understand NVDA internals.

**Comment by seanbudd** [comment]
>I don't have a braille display so can't add or test the state in braille, so if someone could that'd be great

In the future, testing using [Braille Viewer](https://www.nvaccess.org/files/nvda/documentation/userGuide.html#BrailleViewer) should be sufficient for testing.

File: source/NVDAObjects/UIA/__init__.py [hunk 1]
```diff
@@ -1647,6 +1647,11 @@ def _get_states(self):
 		if s!=UIAHandler.handler.reservedNotSupportedValue:
 			if not role:
 				role=self.role
+			if s == UIAHandler.ToggleState_Indeterminate:
+				if role == controlTypes.Role.TOGGLEBUTTON:
+					states.add(controlTypes.State.HALFPRESSED)
+				else:
```
Comment (seanbudd) [inline source/NVDAObjects/UIA/__init__.py:1653]: Shouldn't this explicitly check if this is a checkbox?

```suggestion
				elif role == cpmtrp;Types.Role.CHECKBOX:
```

File: source/controlTypes/deprecatedAliases.py [hunk 2]
```diff
@@ -199,6 +199,7 @@
 STATE_PRESSED = State.PRESSED
 STATE_CHECKED = State.CHECKED
 STATE_HALFCHECKED = State.HALFCHECKED
+STATE_HALFPRESSED = State.HALFPRESSED
```
Comment (seanbudd) [inline source/controlTypes/deprecatedAliases.py:202]: As per the warning in this file, this module is for deprecated code. please do not add to it

File: source/controlTypes/state.py [hunk 3]
```diff
@@ -96,6 +96,7 @@ def negativeDisplayString(self) -> str:
 	# indeterminate progress bar, aka busy indicator. No specific state label.
 	# when combined with role of 'progress bar', role is mutated to 'busy indicator'
 	INDETERMINATE = setBit(44)
+	HALFPRESSED = setBit(45)
```
Comment (seanbudd) [inline source/controlTypes/state.py:99]: I think this name would read better if it had an underscore breaking it up

```suggestion
	HALF_PRESSED = setBit(45)
```

File: source/NVDAObjects/UIA/__init__.py (hunk 1 above)
Comment (mazen428) [reply source/NVDAObjects/UIA/__init__.py:1653]: There are checkable tree view items, so just in case, if it has a toggle pattern then the state is added

**Review by seanbudd — APPROVED** [review]
Thanks @mazen428!

**Pull request merged**